chainer/links/connection/lstm.py

```diff
@@ -197,6 +211,11 @@ def __call__(self, x):
             ~chainer.Variable: Outputs of updated LSTM units.
 
         """
+        if self.upward.has_uninitialized_params:
+            in_size = x.size // len(x.data)
+            self.upward._initialize_params(in_size)
```

Comment:
I agree with keeping the method private as it is.

> Perhaps another option to avoid calling this private method of Linear link would be to change the LSTM link so that it no longer uses Linear link at all and instead just calls the linear function directly.

As `upward` and `lateral` links are [documented](https://github.com/pfnet/chainer/blob/master/chainer/links/connection/lstm.py#L133), we must keep them to support compatibility. Further, if `LSTM` link would hold weight matrices as parameters, we would have to change how to access them (e.g. from `LSTM.upward.W` to `LSTM.upward_W` etc). Although such a compatibility is not supported, it could break some users' codes. So, we need to elaborate the implementation anyway.